File: tournaments/treeDiameter/treeDiameter.py
def treeDiameter(n, tree):
    if not tree:
        return 0

    d = {}
    for t in tree:
        if t[0] not in d:
            d[t[0]] = []
        if t[1] not in d:
            d[t[1]] = []
        d[t[1]].append(t[0])
        d[t[0]].append(t[1])


    def search(n):
        s = set([n])
        stack = [(n,0)]
        r = (n,0)
        while stack:
            c,depth = stack.pop()
            if depth > r[1]:
                r = (c,depth)
            for x in d[c]:
                if x not in s:
                    stack.append((x,depth+1))
                s.add(x)
        return r

    longest = search(tree[0][0])

    return search(longest[0])[1]

File: tournaments/treeDiameter/treeDiameter.js
function treeDiameter(n, tree) {
    f = (x, y) => {
        p = a[y].p
        if (p > -1) {
            f(y, p)
            a[p].c.splice(a[p].c.indexOf(y), 1)
            a[y].c.push(p)
        }
        a[y].p = x
        a[x].c.push(y)
    }
    a = []
    for ([x, y] of tree) {
        if (!a[x]) [x, y] = [y, x]
        if (a[y]) {
            f(x, y)
        } else {
            a[x]
                ? a[x].c.push(y)
                : a[x] = {p: -1, c: [y]}
            a[y] = {p: x, c: []}
        }
    }
    
    f = i => {
        if (i > -1) {
            [x, y] = a[i].c.map(v => f(v)).sort((a, b) => b - a)
            if (~~x + ~~y > m) m = ~~x + ~~y
            return ~~x + 1
        }
    }
    m = 0
    
    if (n) f(a.findIndex(v => v.p == -1))
    
    return m
}
